### PR TITLE
RFC py: Support builtin next() with 2 arguments

### PR DESCRIPTION
**Pull request by stinos**
This is a pretty convenient oneliner (to replace 4 lines otherwise) which I miss in uPy. Couple of questions:
- made it optional with MICROPY_PY_BUILTINS_NEXT2, is that name ok, or should that just go under MICROPY_CPYTHON_COMPAT?
- mp_builtin_next is still used for implementing ```__next__``` with the side effect that ```iter([]).__next__(1)``` now suddenly works, not sure if this is acceptable? And if so this should this be in a seperate cpydiff test probably?
- ok to enable this by default for unix/windows?

### CODE REVIEW

**Comment by pfalcon** [comment]
Why, so easy? May only wonder why nobody did it like that before... Perhaps, this implementation is buggy?

**Comment by pfalcon** [comment]
Heh: https://github.com/micropython/micropython/pull/2266

**Comment by stinos** [comment]
> Heh: #2266

erm, oops, long term memory failure :P

Anyway it looks like my previous implementation was in fact the buggy one because mp_iternext_allow_raise still might raise and that's not covered there.

File: py/modbuiltins.c [hunk 1]
```diff
@@ -316,6 +316,22 @@ MP_DEFINE_CONST_FUN_OBJ_KW(mp_builtin_min_obj, 1, mp_builtin_min);
 
 #endif
 
+#if MICROPY_PY_BUILTINS_NEXT2
+STATIC mp_obj_t mp_builtin_next(size_t n_args, const mp_obj_t *args) {
+    if(n_args == 1) {
```
Comment (pfalcon) [inline py/modbuiltins.c]: Codestyle error.

**Comment by pfalcon** [comment]
Ah, ok, so you actually used mp_iternext() in the case of 2 args, and that's the reason for the "if" in there... Well, workability should be proven by tests. See e.g. gen_yield_from_ducktype.py

 P.S. And yes, we don't have enough feature-focused tests. Tests for some features are spread over various files. That's kinda result of coverage-driven testsuite development. And indeed, increasing testsuite run time is hardly a good idea, it already takes its time to run over a slow connection on a small target. Still, I find some tests are missing, and actually started to add them to my fork.

**Comment by pfalcon** [comment]
Regarding enablement, I'd still recommend to enable it just for micropython-dev and keep production unix binary as is.

**Comment by stinos** [comment]
Fixed missing space and enabled the option in unix coverage build.

Alternative implementation which keeps __next__ behavior the same at the cost of code size: https://github.com/stinos/micropython/commit/ca5a1013dd215b172e46dd0aff8ebbefd919c2fb

Also: the qemu-arm build on Travis fails on this but I cannot reproduce this locally despite having the same compiler version, so no clue what's going on there.

File: tests/cpydiff/builtin_next_arg2.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 """
 categories: Modules,builtins
-description: Second argument to next() is not implemented
+description: Second argument to next() is optional
```
Comment (pfalcon) [inline tests/cpydiff/builtin_next_arg2.py]: The nature of this change is unclear. Yes, second argument to next() is optional. How that is difference to CPython? If this change is merged, this file would need to go completely. (Assuming that  tests/cpydiff/ track differences between the "maximal" MicroPython config and CPython.)

**Review by pfalcon — CHANGES_REQUESTED** [review]
This patch should include a test showing that to-arg next() works with user-defined iterators (classes with `__next__()` method which may raise StopIteration).

**Comment by pfalcon** [comment]
> Also: the qemu-arm build on Travis fails on this but I cannot reproduce this locally despite having the same compiler version, so no clue what's going on there.

Well, Travis output should provide enough info:

~~~
qemu-system-arm -machine integratorcp -cpu cortex-m3 -nographic -monitor null -serial null -semihosting -kernel build/firmware-test.elf > build/console.out
1/584 TESTS FAILED. (67 skipped)
status: 1
...
0.00s$ grep "FAIL" ports/qemu-arm/build/console.out
basics/builtin_next_arg2.py: FAILED
1/584 TESTS FAILED. (67 skipped)
~~~


So, your "skip" logic doesn't really seem to work.

File: tests/cpydiff/builtin_next_arg2.py (hunk 2 above)
Comment (stinos) [reply tests/cpydiff/builtin_next_arg2.py]: >  track differences between the "maximal"

Ok thanks for clearing that out.

**Comment by stinos** [comment]
> This patch should include a test showing that to-arg next() works with user-defined iterators (classes with __next__() method which may raise StopIteration).

Ok. Also added one for user-defined generator function for completeness, anything else maybe?

> So, your "skip" logic doesn't really seem to work.

Indeed, but my question is how to typically debug these Travis-only failures which when it cannot be reproduced locally? Temporarily add something to travis.yml to have it print the .out file or so? Surely someboday has had this problem before?

**Comment by stinos** [comment]
Well, seconds after I wrote the previous command the Travis build now did succeed.

**Comment by dpgeorge** [comment]
Thanks for this.  Merging will wait until after #4356

**Comment by stinos** [comment]
@dpgeorge ok no problem but can you confirm whether we'll have this implemebntation (with the `iter([]).__next__(1)` anomaly) or rather the one from https://github.com/micropython/micropython/compare/master...stinos:next2-alt which is a bit larger in code size but behaves normally?

File: tests/basics/builtin_next_arg2.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+try:
+    print(next(iter([]), 42))
```
Comment (pfalcon) [inline tests/basics/builtin_next_arg2.py]: Why print here?

**Comment by pfalcon** [comment]
Thanks for adding user object test.

> mp_builtin_next is still used for implementing __next__ with the side effect that iter([]).__next__(1) now suddenly works, not sure if this is acceptable? 

That's not ideal.

> And if so this should this be in a seperate cpydiff test probably?

No, if it should be at all, it should be undocumented feature, which nobody ever will guess, and which may be gone any time.

File: tests/basics/builtin_next_arg2.py (hunk 3 above)
Comment (stinos) [reply tests/basics/builtin_next_arg2.py]: So it also serves as a test case. But it's maybe better to move that with the rest indeed, also it's not *that* different from the iter(range(0)) test so perhaps not needed as actual test at all?

**Comment by stinos** [comment]
> That's not ideal.

No it's not. I'm actually +1 for the patch but I'll wait to see if someone else wants to comment on this.

File: tests/basics/builtin_next_arg2.py (hunk 3 above)
Comment (pfalcon) [reply tests/basics/builtin_next_arg2.py]: So, my point is that - if you test for exception, then print() call (and only print() call) is not really needed, and produces code/output noise. It's pretty minor issue of course.

**Comment by dpgeorge** [comment]
> can you confirm whether we'll have this implemebntation (with the `iter([]).__next__(1)` anomaly) or rather the one from [master...stinos:next2-alt](https://github.com/micropython/micropython/compare/master...stinos:next2-alt) which is a bit larger in code size but behaves normally?

Considering that the patch here (with the anomaly) doesn't crash, I think it's fair to do it this way in order to reduce code size (and it's more common to use `next(x)` rather than `x.__next__()` so it's unlikely it'll ever be accidentally found).

**Comment by dpgeorge** [comment]
Ok, merged in 42863830be2d19c7dcdf7ccf1fa66168b1bdc13a with slightly modified test (no longer prints as part of the feature check, and reordered some parts of the body of the test).

----

>     * ok to enable this by default for unix/windows?

I'll probably enable it on stm32, and would be happy to enable it on unix and windows.

> If this change is merged, this file [cpydiff/builtin_next_arg2.py] would need to go completely. (Assuming that tests/cpydiff/ track differences between the "maximal" MicroPython config and CPython.)

This cpydiff file would new need to be removed since the feature is implemented.  But 1) it'd be nice to clarify exactly what cpydiff tracks differences between, it's a subtle issue; and 2) it'd be a shame to lose the workaround encoded by this cpydiff, since the workaround is still useful by ports that don't support next with 2 arguments.

I'm not sure at this point how to resolve these 2 points.

File: tests/basics/builtin_next_arg2.py (hunk 3 above)
Comment (stinos) [reply tests/basics/builtin_next_arg2.py]: Yeah but details matter and without the print is better indeed.

**Comment by stinos** [comment]
Thanks!

**Pull request closed without merging**